crates/runtime/src/virtual_machine.rs

```diff
@@ -335,99 +306,90 @@ impl VirtualMachine {
+                let index = self.state.current_options.len();
                 let node_name = instruction.read_operand(1);
                 // ## Implementation note:
                 // The original calculates the ID in the `ShowOptions` opcode,
                 // but this way is cleaner because it allows us to store a `DialogueOption` instead of a bunch of values in a big tuple.
-                self.state_mut().current_options.push(DialogueOption {
+                self.state.current_options.push(DialogueOption {
                     line,
                     id: OptionId(index),
                     destination_node: node_name,
                     is_available: line_condition_passed,
                 });
-                self.state_mut().program_counter += 1;
+                self.state.program_counter += 1;
             }
             OpCode::ShowOptions => {
                 // If we have no options to show, immediately stop.
-                if self.state().current_options.is_empty() {
-                    *self.execution_state_mut() = ExecutionState::Stopped;
-                    if let Some(dialogue_complete_handler) = &mut self.dialogue_complete_handler {
-                        dialogue_complete_handler.call(&mut self.handler_safe_dialogue);
-                    }
-                    self.state_mut().program_counter += 1;
+                if self.state.current_options.is_empty() {
+                    self.events.push(DialogueEvent::DialogueComplete);
+                    self.state.program_counter += 1;
                     return;
                 }
 
                 // We can't continue until our client tell us which option to pick
-                *self.execution_state_mut() = ExecutionState::WaitingOnOptionSelection;
+                self.set_execution_state(ExecutionState::WaitingOnOptionSelection);
 
                 // Pass the options set to the client, as well as a
                 // delegate for them to call when the user has made
                 // a selection
-                let current_options = self.state().current_options.clone();
-                self.options_handler
-                    .call(current_options, &mut self.handler_safe_dialogue);
-
-                if *self.execution_state() == ExecutionState::WaitingForContinue {
-                    // we are no longer waiting on an option
-                    // selection - the options handler must have
-                    // called SetSelectedOption! Continue running
-                    // immediately.
-                    *self.execution_state_mut() = ExecutionState::Running;
-                }
-                self.state_mut().program_counter += 1;
+                let current_options = self.state.current_options.clone();
+                self.events.push(DialogueEvent::Options(current_options));
+
+                // Implementation note:
+                // Not checking the execution state now since we have no line handler to call `continue_` from.
```

Comment:
If we change the function name, we also need to adjust this comment
```suggestion
                // Not checking the execution state now since we have no line handler to call `r#continue` from.
```

Reply:
[`e885283` (#103)](https://github.com/yarn-slinger/yarn-slinger/pull/103/commits/e885283a9c369a29a3cf3c315a5689c214a54059)